### ISSUE TITLE
Pitest doesn't ignore top for-each loop with a second nested for-each loop

### ISSUE THREAD

**Issue by rnveach** [hcoles/pitest#600]
Based on https://stackoverflow.com/questions/38607129/mutation-testing-negated-conditional-on-a-for-each-loop and https://github.com/hcoles/pitest/issues/286 ,
>  as of 1.2.5 pitest should avoid mutating the conditional in for each loops.

I am finding some weird cases where pitest is mutating for-each loops in Eclipse, and only on certain code, and where it is not mutating them in ant. According to the other issues, pitest shouldn't be causing mutations in these codes.
I only noticed this issue because my group is trying to create a CI for our project and when running pitest through ant (we can't support maven yet) pitest was failing because differences in line and mutation counts. Both lines covered and mutations decreased. When we run our pitest through eclipse everything passes. The differences turned out to be because Eclipse had mutations for the for-each loops and ant didn't.

I am using pitest 1.4.7 with Eclipse Oxygen 4.7.3 and ant 1.9.13 .

Example Class where I produced it locally. I can try to provide a project if needed.
````
package project;

import java.util.ArrayList;
import java.util.List;

public class Example {
	public static List<String> method(List<String> list, int a) {
		final List<String> result = new ArrayList<String>();

		if (list != null) {
			for (String s : list) {

				for (String s2 : list) {
				}
			}
		}

		return result;
	}
}
````

When running pitest through Eclipse, I got 2 mutations on line 11
> negated conditional
> removed call to java/util/Iterator::hasNext

and no mutations on line 13. I would expect both to behave the same. When running through ant, I got no mutations on line 11 at all.

I am not sure if this will fully fix my situation. It is weird ant has no problems generating byte code that pitest will ignore but Eclipse doesn't.

**Comment by rnveach** [comment]
The byte code I am seeing is from eclipse is:
````
    public static List method(List list, int a)
    {
    //    0    0:new             #18  <Class ArrayList>
    //    1    3:dup             
    //    2    4:invokespecial   #20  <Method void ArrayList()>
    //    3    7:astore_2        
    //    4    8:aload_0         
    //    5    9:ifnull          77
    //    6   12:aload_0         
    //    7   13:invokeinterface #21  <Method Iterator List.iterator()>
    //    8   18:astore          4
    //    9   20:goto            67
    //   10   23:aload           4
    //   11   25:invokeinterface #27  <Method Object Iterator.next()>
    //   12   30:checkcast       #33  <Class String>
    //   13   33:astore_3        
    //   14   34:aload_0         
    //   15   35:invokeinterface #21  <Method Iterator List.iterator()>
    //   16   40:astore          6
    //   17   42:goto            57
    //   18   45:aload           6
    //   19   47:invokeinterface #27  <Method Object Iterator.next()>
    //   20   52:checkcast       #33  <Class String>
    //   21   55:astore          5
    //   22   57:aload           6
    //   23   59:invokeinterface #35  <Method boolean Iterator.hasNext()>
    //   24   64:ifne            45
    //   25   67:aload           4
    //   26   69:invokeinterface #35  <Method boolean Iterator.hasNext()>
    //   27   74:ifne            23
    //   28   77:aload_2         
    //   29   78:areturn         
    }
````

**Comment by hcoles** [comment]
Thanks for the detailed report. 

As you've discovered the bytecode produced by the eclipse compiler does not match that produced by javac, which can cause different output for pitest.

If you'd like a consistency between the build and ide it's possible to configure maven (and I therefore assume also ant) to use the eclipse compiler.

The filter which removes for loop mutations is essentially a regex against bytecode. I've encountered various issues getting the filters to match the output of both eclipse and javac, and it looks like you've found another case. 

I'll take a look when I get chance, but unfortunately I'm very short on time. If you'd like to have a shot yourself the code lives here

https://github.com/hcoles/pitest/blob/master/pitest-entry/src/main/java/org/pitest/mutationtest/build/intercept/javafeatures/ForEachLoopFilter.java

**Comment by rnveach** [comment]
> If you'd like a consistency between the build and ide it's possible to configure maven (and I therefore assume also ant) to use the eclipse compiler.

Yes, somehow this didn't come to mind when trying to resolve the issue. This fixes nearly all of the issues in the differences except 1. I am trying to look into why code coverage lines was lowered for 1 class but the mutation score stayed the same. I think the main issue is the eclipse compiler I found isn't the same as my local.

> If you'd like to have a shot yourself

I will try to find time to take a look and see if I can do it.
